hive-metastore/src/main/java/org/apache/iceberg/hive/HiveClientPool.java

```diff
@@ -83,6 +83,20 @@ protected HiveMetaStoreClient reconnect(HiveMetaStoreClient client) {
     return client;
   }
 
+  @Override
+  protected boolean failureDetection(HiveMetaStoreClient client, Exception ex) {
```

Comment:
I might not understand the fix. From reading #1994, I thought the problem was that some client calls throw the wrong exception, a `MetaException` that wraps `TTransportException` instead of `TTransportException`. That causes the reconnect detection to fail and the `MetaException` is thrown instead.

I would then expect the fix to be to detect the `MetaException` that wraps `TTransportException` and return true because the connection should be reconnected.

The implementation here appears to be to try to use the connection again and return true if it fails with `TTransportException` the second time or false (do not reconnect) if it succeeds. But why issue the second call instead of detecting the failure case using `ex.getCause()`?

Reply:
Personally, I support this solution which  is to detect the MetaException that wraps TTransportException and return true.

If we can reach an agreement, I will do it in this way.

Reply:
@pvary, what do you think?

Reply:
MetaException does not have an embedded Exception (no `ex.getCause()` because the MetaException is declared on the Thrift interface and only has a message field) 😢. Also it is used for signaling connection problems and input data issues as well, and the only difference between those is the exception message text. So we can only check the message text to decide on the cause which is too brittle for my taste.

So I would prefer a solution where we do not try to differentiate between the MetaExceptions parsing the text message.

Since we can not have a definite answer when the error is caused by a connection problem or a bad parametrization of the HMS call, I think reconnection every time when we have a MetaException should be avoided if possible. I would use a cheap probe to check the liveliness of the connection (for example `client.getServerVersion()`)

I see the following possibilities:
- Periodic check for the connection - Fix number of extra HMS calls, no extra delay on other HMS calls
- Check when there is a MetaException - Double HMS calls for every call where the response is MetaException 
- Some combination of the above

The decision should depend on the use-cases we would like to cover.
- Who will use the `ClientPool`?
- How likely that we issue calls which are badly parametrized?
- How likely that we will have connection issues?

To answer the questions from the Hive side:
- I do not see places where we can have MetaException because of the wrong parametrization of the calls (if so we have to fix it 😄)
- If there is a MetaException then that is because of the connection issues

OTOH I consider ClientPool as an API which could/should be used by other components, and such an API should handle errors better. I would accept the extra complexity to avoid extra `reconnect` attempts, and even maybe extra HMS calls too, but can be convinced otherwise.

That's my 2 cents.
Peter

Reply:
Hm. What is the exception message when the connection is to blame? I don't like that solution either, but it seems better than making a second RPC call if we have something that is reliable there.

Reply:
The message of `TTransportException` wrapped by `MetaException` is as follows:
```
ERROR - Got exception: org.apache.thrift.transport.TTransportException null 
org.apache.thrift.transport.TTransportException: null at
org.apache.thrift.transport.TIOStreamTransport.read(TIOStreamTransport.java:132) ~[libthrift-0.9.3.jar:0.9.3] at
org.apache.thrift.transport.TTransport.readAll(TTransport.java:86) ~[libthrift-0.9.3.jar:0.9.3] at
org.apache.thrift.protocol.TBinaryProtocol.readAll(TBinaryProtocol.java:429) ~[libthrift-0.9.3.jar:0.9.3] at
org.apache.thrift.protocol.TBinaryProtocol.readI32(TBinaryProtocol.java:318) ~[libthrift-0.9.3.jar:0.9.3] at
org.apache.thrift.protocol.TBinaryProtocol.readMessageBegin(TBinaryProtocol.java:219) ~[libthrift-0.9.3.jar:0.9.3] at
org.apache.thrift.TServiceClient.receiveBase(TServiceClient.java:77) ~[libthrift-0.9.3.jar:0.9.3] at
org.apache.hadoop.hive.metastore.api.ThriftHiveMetastore$Client.recv_get_all_tables(ThriftHiveMetastore.java:1487) ~[hive-metastore-2.3.7.jar:2.3.7] at
org.apache.hadoop.hive.metastore.api.ThriftHiveMetastore$Client.get_all_tables(ThriftHiveMetastore.java:1474) ~[hive-metastore-2.3.7.jar:2.3.7] at
org.apache.hadoop.hive.metastore.HiveMetaStoreClient.getAllTables(HiveMetaStoreClient.java:1451) [hive-metastore-2.3.7.jar:2.3.7] 
```

Reply:
That seems pretty specific to me: `Got exception: org.apache.thrift.transport.TTransportException`. @pvary, would you be okay with matching that string?

Reply:
@rdblue, @Jecarm: Moving forward with this string solves this specific problem. Later we can still reconsider if we find other cases.

Thanks @Jecarm for the investigations!

Reply:
I have submitted the new code. Please review it. @pvary @rdblue